chainstate/test-suite/src/tests/block_status.rs

```diff
@@ -13,202 +13,488 @@
+        let block1_id = block1.get_id();
+        tf.process_block(block1, BlockSource::Local).unwrap();
 
+        let block1_status = get_block_status(&tf, &block1_id);
+        assert_ok_block_at_stage(block1_status, BlockValidationStage::CheckBlockOk);
+
+        // In the second block, create a tx that tries to spend burnt coins.
+        let (block2_id, result) = process_block_spend_tx(
+            &mut tf,
+            &block1_id.into(),
+            &tx1.transaction().get_id(),
+            &mut rng,
+        );
+        assert!(result.unwrap().is_none());
         let block2_status = get_block_status(&tf, &block2_id);
-        assert_eq!(
-            block2_status.last_valid_stage(),
-            BlockValidationStage::CheckBlockOk
+        assert_ok_block_at_stage(block2_status, BlockValidationStage::CheckBlockOk);
+
+        // On top of block 2 create a subtree of good blocks.
+
+        let (block3_id, result) = process_block(&mut tf, &block2_id.into(), &mut rng);
+        assert!(result.unwrap().is_none());
+        let block3_status = get_block_status(&tf, &block3_id);
+        assert_ok_block_at_stage(block3_status, BlockValidationStage::CheckBlockOk);
+
+        let (block4_id, result) = process_block(&mut tf, &block3_id.into(), &mut rng);
+        assert!(result.unwrap().is_none());
+        let block4_status = get_block_status(&tf, &block4_id);
+        assert_ok_block_at_stage(block4_status, BlockValidationStage::CheckBlockOk);
+
+        // Note: using block3 as the parent is intended.
+        let (block5_id, result) = process_block(&mut tf, &block3_id.into(), &mut rng);
+        assert!(result.unwrap().is_none());
+        let block5_status = get_block_status(&tf, &block5_id);
+        assert_ok_block_at_stage(block5_status, BlockValidationStage::CheckBlockOk);
+
+        // No reorg has occurred yet.
+        assert!(!tf.is_block_in_main_chain(&block1_id));
+        assert!(!tf.is_block_in_main_chain(&block2_id));
+        assert!(!tf.is_block_in_main_chain(&block3_id));
+        assert!(!tf.is_block_in_main_chain(&block4_id));
+        assert!(!tf.is_block_in_main_chain(&block5_id));
+
+        assert_block_data_exists(&tf, &block1_id, true);
+        assert_block_data_exists(&tf, &block2_id, true);
+        assert_block_data_exists(&tf, &block3_id, true);
+        assert_block_data_exists(&tf, &block4_id, true);
+        assert_block_data_exists(&tf, &block5_id, true);
+
+        // This will trigger a reorg that should fail.
+        tf.create_chain(&block4_id.into(), 2, &mut rng).unwrap_err();
+
+        // FIXME: currently, successful validation that occurred during a failed reorg is not
+        // recorded. Though it's possible to make "reorganize" return the start of the new chain
+        // on error (or to refactor it to accept a new chain), the idea of marking certain blocks
+        // as fully valid after a reorg has failed might look error prone. Is it worth implementing it?
```

Comment:
I wouldn't care. The only issue with this is an attack vector for large reorgs, but then since peers get banned, it doesn't really matter.